File: Server Apps/Example App/ExampleApp.py
from RPAE import RpaeApp


class ExampleApp(RpaeApp):
    def __init__(self):
        # set the name of your app configuration, usually app.ini. default is 'app.ini'
        super().__init__()
        # super().__init__(configFile='myconfig.ini')

    def onMessage(self, message, origin):
        print("new message from uid:", origin['uid'], "message:", message)

    # If there is an error it will already be printed so add any other stuff you want to do here
    def onError(self, error):
        pass

    # executes when the connection is made and the handshake was successful
    def onOpen(self):
        print('available peers: ', self.peers)
        print("my app ID:", self.appID)
        print("my uid:", self.uid)
        print("my location:", self.location)
        print("number of connected peers", len(self.peers))
        print("connected with:", self.host)
        print("app type:", self.appType)
        print("peer address:", self.peerAddress)
        print("peer name:", self.peerName)
        print("connected as:", self.peerOrigin)
        print("isReady is true once the handshake is done.", self.isReady)
        print("Apptypes contains targets one can send messages too", self.appTypes)
        # lets send hello to all the webclients!
        self.sendMessage(message="hello to all clientApps!")
        # it's a bit sad if we left anyone out. (self.appTypes[0] contains 'all')
        self.sendMessage(message="hello to all!", appType=self.appTypes[0])  # default app type is: clientApp

    # executes when a new clients is connected
    def onNewPeer(self, peer):
        print("new peer connected with uid:", peer.uid, "app type:", peer.appType)
        print("lets say hello to our new friend!")
        self.sendMessage(message='hello new peer!!', appType='uid', uidTarget=peer.uid)

    def onPeerDisconnected(self, peer):
        print("Bye peer! uid:", peer.uid, "app type:", peer.appType)

    def init(self):
        print("init is called when the start function triggers.")
        print("This function triggers before the handshake is done. as you can see connected is", self.connected)


# The main function to start the app
if __name__ == "__main__":
    app = ExampleApp()      # First create an object of your app
    ExampleApp().start()    # Connect to the server and start listening

File: Webclient/apps/ExampleApp/js/ExampleApp.js
class ExampleApp extends RpaeApp{
    constructor(){
        super('ExampleApp');                    // Set your AppID here
        this.host = 'ws://192.168.2.8:9738';    // Setting host will immediately start up the connection.
    }

    onOpen(data){
        console.log('Available peers: ', this.peers);
        console.log('My app ID:', this.appID);
        console.log('My unique ID:', this.uid); // uid is and unique id given by the server.
        console.log('Number of connected peers:', Object.keys(this.peers).length);
        console.log('Connected with:', this.host);
        console.log('My app type:', this.appType);
        console.log('Peer address:', this.peerAddress);
        console.log('Peer name', this.peerName);
        console.log('Connected as:', this.peerOrigin);
        console.log('isReady will be true once the connection is established AND had an successful handshake', this.isReady);
        console.log('The apptypes usuable for sendMessage:', this.appTypes);
        // let's send a message to all server apps!
        this.sendMessage('hello from webclient example app!');
        //it's a bit sad if we left anyone out. (self.appTypes[0] contains 'all')
        this.sendMessage("hello to all!", this.appTypes[0]);  // default app type is: serverApp
    }

    onMessage(message, origin){
        console.log(message, origin);
    }

    onClose(data){
        console.log('Websocket closed with data: ', data);
    }

    onError(error){
        console.log('An error has occured when trying to connect.. ', error);
    }

    onNewPeer(peer){
        console.log('New peer connected with uid:', peer.uid, 'app type:', peer.appType);
        console.log('Lets say hello to our new friend!');
        this.sendMessage('Hello friend!', 'uid', peer.uid);
    }

    onPeerDisconnect(peer){
        console.log('Bye peer!', peer);
    }
}